src/pay/components/PayProvider.tsx

```diff
@@ -0,0 +1,281 @@
+  const [errorMessage, setErrorMessage] = useState<string>('');
+
+  // Component lifecycle
+  const { lifecycleStatus, updateLifecycleStatus } = useLifecycleStatus({
+    statusName: PAY_LIFECYCLESTATUS.INIT,
+    statusData: {},
+  });
+
+  // Transaction hooks
+  const fetchContracts = useCommerceContracts({
+    chargeHandler,
+    productId,
+  });
+
+  const { status, writeContractsAsync } = useWriteContracts({
+    /* v8 ignore start */
+    mutation: {
+      onSuccess: (id) => {
+        setTransactionId(id);
+      },
+      onError: (e) => {
+        const errorMessage = isUserRejectedRequestError(e)
+          ? 'Request denied.'
+          : GENERIC_ERROR_MESSAGE;
+        updateLifecycleStatus({
+          statusName: PAY_LIFECYCLESTATUS.ERROR,
+          statusData: {
+            code: 'PmUWCSh01',
+            error: e.message,
+            message: errorMessage,
+          },
+        });
+      },
+    },
+    /* v8 ignore stop */
+  });
+  const { data } = useCallsStatus({
+    id: transactionId,
+    query: {
+      /* v8 ignore next 3 */
+      refetchInterval: (query) => {
+        return query.state.data?.status === 'CONFIRMED' ? false : 1000;
+      },
+      enabled: !!transactionId,
+    },
+  });
+  const transactionHash = data?.receipts?.[0]?.transactionHash;
+  const { data: receipt } = useWaitForTransactionReceipt({
+    hash: transactionHash,
+  });
+
+  // Component lifecycle emitters
+  useEffect(() => {
+    onStatus?.(lifecycleStatus);
+  }, [
+    lifecycleStatus,
+    lifecycleStatus.statusData, // Keep statusData, so that the effect runs when it changes
+    lifecycleStatus.statusName, // Keep statusName, so that the effect runs when it changes
+    onStatus,
+  ]);
+
+  // Set transaction pending status when writeContracts is pending
+  useEffect(() => {
+    if (status === 'pending') {
+      updateLifecycleStatus({
+        statusName: PAY_LIFECYCLESTATUS.PENDING,
+        statusData: {},
+      });
+    }
+  }, [status, updateLifecycleStatus]);
+
+  // Trigger success status when receipt is generated by useWaitForTransactionReceipt
+  useEffect(() => {
+    if (!receipt) {
+      return;
+    }
+    updateLifecycleStatus({
+      statusName: PAY_LIFECYCLESTATUS.SUCCESS,
+      statusData: {
+        transactionReceipts: [receipt],
+        chargeId: chargeId,
+        receiptUrl: `https://commerce.coinbase.com/pay/${chargeId}/receipt`,
+      },
+    });
+  }, [chargeId, receipt, updateLifecycleStatus]);
+
+  // biome-ignore lint/complexity/noExcessiveCognitiveComplexity: TODO Refactor this component to deprecate funding flow
+  const handleSubmit = useCallback(async () => {
+    try {
+      // Open Coinbase Commerce receipt
+      if (lifecycleStatus.statusName === PAY_LIFECYCLESTATUS.SUCCESS) {
+        window.open(
+          `https://commerce.coinbase.com/pay/${chargeId}/receipt`,
+          '_blank',
+          'noopener,noreferrer',
+        );
+        return;
+      }
```

Comment:
may be more straightforward to have separate callbacks for handleFundingFlow and handleViewReceipt and then call different callbacks in the button based on the status

Reply:
I'll refactor this as a fast follow once we have USDC magic spend

Reply:
I don't think it makes sense to add a `handleFundingFlow.ts` just to delete it in a week or so